### PR TITLE
Integrate all (dummy) subsystems with the Overseer

### LINKED ISSUE #1317: Integrate all subsystems with the Overseer
This can speculatively be done before the subsystems are actually implemented.

However, I'd recommend introducing a helper `DummySubsystem` first, which implements `Subsystem` for all message types, creating a future that spins forever. Then users of the Overseer can just supply `DummySubsystem` to those subsystems that it doesn't need to do anything for. This will make test setup dramatically easier.

### PR DESCRIPTION
**Pull request by ordian**
Fixes #1317.

Please bear with me while I make these baby steps.
This boilerplate looks ugly so as an experiment I looked at a type map approach 2e934c3662d7abfd956880f94fd7b0da0be72ce7 (instead of storing a bunch of fields, we can store a `HashMap<Type, Box<SomeTrait>>`) but then reverted this change for now. Let me know if it's worth exploring.

### CODE REVIEW

**Comment by ordian** [comment]
@coriolinus thanks for the detailed feedback! 

I was hesitant to throw macros at the problem, because in my experience while it can solve some problems, it inevitably create others, like code readability (now you need to read the source of the macro/docs/tests), maintenance, longer compile times, IDE-unfriendliness. But given the love of macros at Parity, I might reconsider :)

>  I think that you're also right that the runtime typemap is not the right way to handle it, because it prevents things like exhaustive matching.

Actually, I don't think it does prevent exhaustive matching, but to be fair I haven't explored this avenue fully. It also can guarantee exhaustive listing of subsystem fields via iteration.

**Comment by coriolinus** [comment]
You're right that there are a bunch of downsides to throwing macros at a problem, but they do come with one key upside: they make correctness a lot easier. If we use macros set up as I suggested, then adding a new subsystem means that, beyond implementing the subsystem itself, you just need to add it to the `declare_types` list. Without it, you have to add a copy-paste case in a bunch of different places, none of which are easy to grep for. IMO that makes maintenance easier, not harder.

**Comment by ordian** [comment]
@coriolinus makes a good point about correctness though. It's not currently guaranteed at compile time that we didn't forget to check a certain subsystem e.g. in `broadcast_signal`. OTOH, the tests should catch it and I don't know if we we'll be adding any new subsystem anytime soon.

File: node/overseer/src/lib.rs [hunk 1]
```diff
@@ -457,8 +478,16 @@ where
 	/// ```
 	pub fn new(
 		leaves: impl IntoIterator<Item = BlockInfo>,
-		validation: impl Subsystem<OverseerSubsystemContext<CandidateValidationMessage>> + Send,
+		candidate_validation: impl Subsystem<OverseerSubsystemContext<CandidateValidationMessage>> + Send,
 		candidate_backing: impl Subsystem<OverseerSubsystemContext<CandidateBackingMessage>> + Send,
+		candidate_selection: impl Subsystem<OverseerSubsystemContext<CandidateSelectionMessage>> + Send,
```
Comment (rphmeier) [inline node/overseer/src/lib.rs:483]: Perhaps this should be refactored out to a separate struct

Comment (ordian) [reply node/overseer/src/lib.rs]: Not sure if 5e9d6699ad1113a72ef302be8afe96da33721c87 is what you meant

File: node/overseer/src/lib.rs [hunk 2]
```diff
@@ -375,22 +375,58 @@ pub struct Overseer<S: Spawn> {
...
+/// As any entity that satisfies the interface may act as a [`Subsystem`] this allows
+/// mocking in the test code:
+///
+/// Each [`Subsystem`] is supposed to implement some interface that is generic over
+/// message type that is specific to this [`Subsystem`]. At the moment not all
+/// subsystems are implemented and the rest can be mocked with the [`DummySubsystem`].
+///
+/// [`Subsystem`]: trait.Subsystem.html
+/// [`DummySubsystem`]: struct.DummySubsystem.html
+pub struct AllSubsystems<CV, CB, CS, SD, AD, BD, P, RA, AS, NB>
```
Comment (rphmeier) [inline node/overseer/src/lib.rs:389]: 🙏

File: node/overseer/src/lib.rs (hunk 1 above)
Comment (rphmeier) [reply node/overseer/src/lib.rs]: Exactly what I meant

**Pull request merged**